File: Turquoise/UI/TQHeaderDownloadProgressView.m
#import "TQHeaderDownloadProgressView.h"

#import "TQNNTPGroup.h"
#import "UIView+NibLoader.h"

@implementation TQHeaderDownloadProgressView {
  NSString *_groupId;

  IBOutlet UIActivityIndicatorView *_progressIndicator;
  IBOutlet UILabel *_infoLabel;
  IBOutlet UILabel *_progressLabel;
}

- (instancetype)initWithGroupId:(NSString *)groupId {
  self = [self init];
  if (self) {
    _groupId = [groupId copy];
    _infoLabel.text = [NSString stringWithFormat:@"Downloading headers for group:\n'%@'...", _groupId];
    _progressLabel.text = @"0%";
    _progressIndicator.transform = CGAffineTransformMakeScale(1.5, 1.5);

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(progressDidUpdate:)
                                                 name:kHeaderDownloadProgressNotification
                                               object:nil];
  }
  return self;
}

- (instancetype)init {
  self = (TQHeaderDownloadProgressView *)[UIView tq_loadFromNib:@"TQHeaderDownloadProgressView" owner:self];
  return self;
}

- (void)progressDidUpdate:(NSNotification *)notification {
  NSNumber *progress = notification.userInfo[kHeaderDownloadProgressAmountKey];
  if (progress) {
    _progressLabel.text = [NSString stringWithFormat:@"%@%%", progress];
  }
}

@end

File: Turquoise/UI/TQHeaderDownloadProgressView.swift
import Foundation
import UIKit

class TQHeaderDownloadProgressView : UIView {
  var groupId: String? {
    didSet {
      var text = "Downloading headers"
      if let groupId = groupId {
        text += " for group:\n'\(groupId)'..."
      }
      self.infoLabel.text = text
    }
  }

  @IBOutlet var progressIndicator: UIActivityIndicatorView!
  @IBOutlet var infoLabel: UILabel!
  @IBOutlet var progressLabel: UILabel!

  class func loadFromNib(groupId: String) -> TQHeaderDownloadProgressView? {
    guard let view = UIView.tq_load(from: "TQHeaderDownloadProgressView", owner: self) as? TQHeaderDownloadProgressView else {
      return nil
    }

    view.groupId = groupId
    view.progressLabel.text = "0%"
    view.progressIndicator.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
    NotificationCenter.default.addObserver(view,
                                           selector: #selector(progressDidUpdate(_:)),
                                           name: TQNNTPGroup.headerDownloadProgressNotification,
                                           object: nil)
    return view
  }

  func progressDidUpdate(_ notification: Notification) {
    if let progress = notification.userInfo?[TQNNTPGroup.headerDownloadProgressAmountKey] {
      self.progressLabel.text = "\(progress)%"
    }
  }
}
